File: Module 07/Lab/StarWars_API/src/main/java/edu/neumont/ApiBase.java
package edu.neumont;

import edu.neumont.dtos.DtoBase;
import edu.neumont.interfaces.IApi;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

public abstract class ApiBase<T extends DtoBase> implements IApi<T>
{
    protected static final String baseUrl = "https://swapi.dev/api/";

    protected String call(String url) throws IOException, InterruptedException
    {
        String theUrl = url.replace("http://", "https://");
        HttpClient client = HttpClient.newHttpClient();

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(theUrl))
                .build();

        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

        return response.body();
    }
}

File: Module 07/dotNet/StarWars.Common/StarWars.Api/ApiBase.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using Newtonsoft.Json;
using StarWars.Api.Models;
using StarWars.Common.Dtos;
using StarWars.Common.interfaces;

namespace StarWars.Api
{
    public abstract class ApiBase<T>: IApi<T> where T: DtoBase
    {
        protected const string BaseUrl = "https://swapi.dev/api/";

        public abstract List<T> GetAll();
        public abstract List<T> GetPage(int pageNumber);
        public abstract List<T> GetPage(string url);
        public abstract T Get(string url);
        public abstract T Get(int id);

        protected SwapiPageResponse<T> GetPageResponse(String url)
        {
            var responseString = CallApi(url);

            if (responseString != null)
            {
                var settings = new JsonSerializerSettings
                               {
                                   ContractResolver = new CustomContractResolver()
                               };

                return JsonConvert.DeserializeObject<SwapiPageResponse<T>>(responseString, settings);
            }

            var response = (SwapiPageResponse<T>) Activator.CreateInstance(typeof(SwapiPageResponse<T>));
            return response;
        }

        protected T GetItemResponse(String url)
        {
            var responseString = CallApi(url);

            var settings = new JsonSerializerSettings
                           {
                               ContractResolver = new CustomContractResolver()
                           };

            if (responseString != null)
            {
                return JsonConvert.DeserializeObject<T>(responseString, settings);
            }

            return null;
        }


        protected string CallApi(string url)
        {
            var client = new HttpClient();
            client.BaseAddress = new Uri(url);

            client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            var response = client.GetAsync(url).Result;

            return response.IsSuccessStatusCode ? response.Content.ReadAsStringAsync().Result : null;
        }
    }
}